source/agora/consensus/EnrollmentManager.d

```diff
@@ -738,6 +771,42 @@ public class EnrollmentManager
         return this.enroll_pool.getEnrollment(enroll_hash);
     }
 
+    /***************************************************************************
+
+        Add missing enrollment information
+
+        Params:
+            enroll_key = the key of the missing enrollment
+
+    ***************************************************************************/
+
+    public void addMissingEnrollment (in Hash enroll_key) @safe nothrow
+    {
+        this.missing_enrolls.put(enroll_key);
+    }
+
+    /***************************************************************************
+
+        Get the UTXO keys of mssing enrollments
+
+        Returns:
+            set of UTXO keys that EnrollmentManager is missing
+
+    ***************************************************************************/
+
+    public Set!Hash getMissingEnrollments () @safe
+    {
+        Set!Hash local_missing_enrolls;
+        foreach (key; this.missing_enrolls)
+        {
+            if (this.getEnrollment(key) == Enrollment.init)
+                local_missing_enrolls.put(key);
+        }
+
+        this.missing_enrolls = local_missing_enrolls;
+        return missing_enrolls;
+    }
```

Comment:
Why the `Set` duplication ?

Reply:
The `missing_enrollments` will be used later so it must have the latest missing enrollments. If you meant another way, please let me know.

Reply:
I am not sure why we need a local `Set` for this. This seems to replicate what was done with "missing transactions", but the problem was different: Missing transactions uses a `Set` because we see the hashes in the SCP statement but we can't match a data structure. On the other hand, we haven't seen the hashes of `Enrollment` that are missing, we don't even know if they exist or not.

Reply:
I see. I think we need to store `Enrollment`s for the missing enrollments instead of hashes as commented on other threads. For `Set` duplication, I didn't know that `remove` function can be used in the iteration. I understand the local variable is not needed now. Thx for pointing out the problem.